kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/ColumnMapping.java

```diff
@@ -354,11 +354,14 @@ private static Optional<Metadata> assignColumnIdAndPhysicalName(
       newSchema = rewriteFieldIdsForIceberg(newSchema, maxColumnId);
     }
 
+    boolean shouldUpdateMaxId =
+        TableConfig.COLUMN_MAPPING_MAX_COLUMN_ID.fromMetadata(metadata) != maxColumnId.get();
+
```

Comment:
I think there's a problem, we should make sure the max ID we choose is the max of what's in the metadata and what is computed. That's important so that later added columns do not impede older but dropped columns.

Reply:
I mean not literally on this line but basically I think it's readable if we extract the max before this line as int largestIdAcrossSchemas = max(fromMetadata, maxcolumnId), and then later on we can ceck if it was equal or not for the case where the schemas are the same.

Reply:
+1

Reply:
So that's literally what we do above, I can maybe rename the variable to be more clear?
```
    AtomicInteger maxColumnId =
        new AtomicInteger(
            Math.max(
                Integer.parseInt(
                    metadata
                        .getConfiguration()
                        .getOrDefault(COLUMN_MAPPING_MAX_COLUMN_ID_KEY, "0")),
                findMaxColumnId(oldSchema)));
```
I think `oldMaxColumnId` is poorly named!

Reply:
Oh yeah I think you're right. it's more than just naming that made this logic confusing. That code snippet you linked, maxColumnId is the max of the metadata and the passed in schema. We then go through a *possibly* new schema, and *possibly* mutate the maxColumnId (only if there are new fields with higher IDs). So the invariant that it's the max across all schema's. 

TLDR It's kinda confusing as it's written because it's all the same maxColumnId just being potentially mutated. I'm fine as it is, we can revisit later but yeah maybe breaking up the logic into 

1. "My current maxColumnId"
2. "The new schema max ID"
3. "The one I write to metadata is max(1, 2)"

makes it easier, or maybe that's just how I'm thinking about it

Reply:
Basically it's checking that the maxColumnId has not changed. Now, by nature of entering this if statement, we know the maxColumnId has not changed or else we would have to update it.

Reply:
Yeah I agree with your general idea of making it more clear. Added some comments for now to hopefully clarify things